### PR TITLE
Optimize gas nodes consumption

### LINKED ISSUE #1212: Optimize gas tree nodes consumption
Currently for each consuming node a `catch_value` procedure is called, which is an iterative procedure. Node consumption is an iterative procedure as well. So algorithm works for O(N^2). We can use internal tree properties in order to reduce the complexity to O(n). For example, if the result of `catch_value` procedure in `consume` is `Caught(non_zero_value)`, there is no reason to call it again for ancestors. Same is with `Missed`.

### PR DESCRIPTION
**Pull request by SabaunT**
Resolves #1212 .

- Makes optional run of `catch_value` procedure depending on the result of the child's run (in `try_remove_consumed_ancestors`).
- Describes all the optimization details in the doc comments.

@gear-tech/dev

### CODE REVIEW

**Review by breathx — APPROVED** [review]

File: common/src/gas_provider/internal.rs [hunk 1]
```diff
@@ -246,18 +246,41 @@ where
...
     /// 2. Also in that case cascade ancestors consumption will last until either the patron node or the first ancestor with specified child is found.
     /// 3. If `catch_value` call ended up with `CatchValueOutput::Caught(x)` in `consume`, all the calls of `catch_value` on ancestor nodes will be `CatchValueOutput::Caught(0)`.
-    pub(super) fn try_remove_consumed_ancestors(
+    /// That's due to the 12-th invariant stated in [`super::property_tests`] module docs. When node becomes consumed without unspec refs (i.e., stops being a patron)
+    /// `consume` procedure call on such node either moves value upstream (if there is an ancestor patron) or returns value to the origin. So any repetitive `catch_value` call on
+    /// such nodes results in `CatchValueOutput::Caught(0)` (if there are is ancestor patron). So if `consume` procedure on the node with `key` id resulted in value being caught,
+    /// it means that there are no ancestor patrons, so non of `cath_value` calls on the node's ancestors will return `CatchValueOutput::Missed`, but will return `CatchValueOutput::Caught(0)`.
```
Comment (ekovalev) [inline common/src/gas_provider/internal.rs:256]: Typo: non of `cath_value` => none of `catch_value`

Comment (SabaunT) [reply common/src/gas_provider/internal.rs]: done

File: common/src/gas_provider/internal.rs [hunk 2]
```diff
@@ -246,18 +246,41 @@ where
...
+    /// will give the same `CatchValueOutput::Missed` result (they all have same ancestor patron, which will receive their values).
     /// 2. Also in that case cascade ancestors consumption will last until either the patron node or the first ancestor with specified child is found.
     /// 3. If `catch_value` call ended up with `CatchValueOutput::Caught(x)` in `consume`, all the calls of `catch_value` on ancestor nodes will be `CatchValueOutput::Caught(0)`.
-    pub(super) fn try_remove_consumed_ancestors(
+    /// That's due to the 12-th invariant stated in [`super::property_tests`] module docs. When node becomes consumed without unspec refs (i.e., stops being a patron)
+    /// `consume` procedure call on such node either moves value upstream (if there is an ancestor patron) or returns value to the origin. So any repetitive `catch_value` call on
+    /// such nodes results in `CatchValueOutput::Caught(0)` (if there are is ancestor patron). So if `consume` procedure on the node with `key` id resulted in value being caught,
```
Comment (gshep) [inline common/src/gas_provider/internal.rs:255]: `... (if there is an ancestor ...`?

File: common/src/gas_provider/internal.rs [hunk 3]
```diff
@@ -246,18 +246,41 @@ where
...
         key: MapKey,
+        descendant_catch_output: CatchValueOutput<Balance>,
     ) -> Result<ConsumeOutput<NegativeImbalance<Balance, TotalValue>, ExternalId>, Error> {
         let mut node_id = key;
         let mut node = Self::get_node(key).ok_or_else(InternalError::node_not_found)?;
         let mut consume_output = None;
         let (_, origin) = Self::get_origin(key)?.expect("existing node always have origin");
 
+        // Descendant's `catch_value` output is used for the sake of optimization.
+        // We could easily run `catch_value` in the bellow `while` loop each time
```
Comment (gshep) [inline common/src/gas_provider/internal.rs:267]: misprint: `in the below ...`

**Review by gshep — APPROVED** [review]
Good work! :+1:

File: common/src/gas_provider/internal.rs (hunk 3 above)
Comment (SabaunT) [reply common/src/gas_provider/internal.rs]: done

**Pull request merged**